Split up content of EmailListTermsConditionsForm within onboarding

vaidehijoshi opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR makes two major changes to the onboarding flow:
1) It moves the terms and conditions/code of conduct into the intro slide.
2) It moves the email preferences into the last slide (before the closing slide, which [will ultimately disappear](https://github.com/thepracticaldev/dev.to/issues/6552)).

It also prepends `v2:` to some of the `updateOnboarding()` functions and some lines to ignore linters where necessary. Also reorganizes the tests.

## Related Tickets & Documents
Closes https://github.com/thepracticaldev/dev.to/issues/6545.
Closes https://github.com/thepracticaldev/dev.to/issues/6550.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

Before this change, our intro slide didn't have much going on inside of it...
<img width="598" alt="Screen Shot 2020-04-14 at 4 55 50 PM" src="https://user-images.githubusercontent.com/6921610/79285027-2bd4ec00-7e71-11ea-8313-6fd50071581c.png">

...and all of the "preference" checkboxes lived in the second slide:
<img width="823" alt="Screen Shot 2020-04-14 at 4 55 58 PM" src="https://user-images.githubusercontent.com/6921610/79285049-3beccb80-7e71-11ea-9c34-e3844162afbe.png">

With this PR, this is now split up into different slides. The terms and conditions and code of conduct appear on the intro slide, and the button is disabled unless both boxes are checked:
<img width="621" alt="Screen_Shot_2020-04-14_at_4_52_31_PM" src="https://user-images.githubusercontent.com/6921610/79285167-9f76f900-7e71-11ea-9378-c38f879ffd9b.png">
<img width="606" alt="Screen Shot 2020-04-14 at 4 52 44 PM" src="https://user-images.githubusercontent.com/6921610/79285186-a7cf3400-7e71-11ea-9108-30417ff7e294.png">

The email preferences are now the second-to-last slide, appearing just before the closing slide:
<img width="828" alt="Screen Shot 2020-04-14 at 4 53 06 PM" src="https://user-images.githubusercontent.com/6921610/79285192-ad2c7e80-7e71-11ea-89a1-2774e9d0a9b0.png">

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

vaidehijoshi commented:
I suggest reviewing this PR commit-by-commit! 😊

vaidehijoshi commented:
I'm tagging @nickytonline + @rhymes + @maestromac here, who I am sure know this code better than I do and might be able to point to what I'm doing wrong. I am genuinely confused about why [these tests are failing](https://travis-ci.com/github/thepracticaldev/dev.to/builds/160306053#L1731-L1766). I spent quite some time trying to debug today and I know that it is due to the [last commit](https://github.com/thepracticaldev/dev.to/pull/7294/commits/67384c45dd5492d7ce47d96340e46d18ab82aba4). I could use some guidance on how I could have caused it to fail in this way... 😕 Any ideas?

vaidehijoshi commented:
I'm going to try to continue debugging why some of these specs are failing, but marking this as ready for review since the feature does actually work for me...hopefully anyone who reviews this can provide some insight into what on earth I'm doing to cause these failures 😖

lisasy commented on app/assets/stylesheets/preact/onboarding-modal.scss:
```diff
@@ -582,6 +594,14 @@ $onboarding-user-selected-hover: rgba(71, 85, 235, 0.2);
   }
 
+  .checkbox-form-wrapper {
+    padding: 32px;
```
since we're using Crayons spacing units above, can we do that here?

According to `_import.scs`s, `32px` can mean `$su-7`

lisasy commented on app/assets/stylesheets/preact/onboarding-modal.scss:
```diff
@@ -582,6 +594,14 @@ $onboarding-user-selected-hover: rgba(71, 85, 235, 0.2);
+    padding: 32px;
+
+    .checkbox-form {
+      padding-bottom: 24px;
```
since we're using Crayons spacing units above, can we do that here?

According to `_import.scs`s, `24px` can mean `$su-6`

vaidehijoshi replied:
Yes! I missed that the definitions for variables are in `_import.scss`. I'll pull from there and update :)

vaidehijoshi replied:
@lisasy Is there any reason to not add a reference link to `_import.scss` to https://dev.to/page/crayons? I think it would have been helpful to know, and it seems good to add :)

nickytonline commented on app/assets/stylesheets/preact/onboarding-modal.scss:
```diff
@@ -109,11 +109,19 @@
   width: 100%;
   border-bottom: 2px solid var(--box-darker);
 
+  &.intro-slide {
```
It's not documented yet, but consider to start using the CSS utility classes that @ludwiczakpawel has created. For reference see https://github.com/thepracticaldev/dev.to/blob/master/app/assets/stylesheets/config/_generator.scss.

vaidehijoshi replied:
@nickytonline Hm, I see what you mean, but without documentation I am not sure how to go about using it. Do you think this could be done in a separate PR, once it has been documented? Alternatively, can you point me to a place where we actually already use it in the app so I can figure out how to apply it?

nickytonline replied:
Fair point about the documentation. Well played, well played. 😄 @ludwiczakpawel, this is probably a good task to do once we hit the next cooldown period.

@vaidehijoshi, it's still fairly new, so it's used in a few places, but only certain utility classes. I think it's fine to do this in a future PR.

vaidehijoshi replied:
I'd love to use and apply the new utility classes; cooldown does seem like a great time for me to practice using them ! 😄

nickytonline commented on app/javascript/onboarding/components/IntroSlide.jsx:
```diff
@@ -2,27 +2,98 @@ import { h, Component } from 'preact';
+    this.setState({ text: document.getElementById(id).innerHTML });
+  }
+
+  buttonIsDisabled() {
```
Consider renaming this.
```suggestion
  isButtonDisabled() {
```

nickytonline commented on app/javascript/onboarding/components/IntroSlide.jsx:
```diff
@@ -2,27 +2,98 @@ import { h, Component } from 'preact';
+    return !checked_code_of_conduct || !checked_terms_and_conditions;
+  }
+
+  backToSlide() {
```
This appears to be used in only one place. It may not merit being a function. Also, what does setting the `text` state property to `false` do exactly when things rerender? It looks likes it closes the terms and conditions?

vaidehijoshi replied:
This was preexisting in the original component I transferred it from...but I can render the function's contents inline since you're right that it's only used in one place.

I actually think, now that I'm looking at this a second time around, that the `text` property should be set to `null` rather than `false`. When it is set to `false` on the state, the component re-`render`s, and the [coc/terms + conditions view](https://github.com/thepracticaldev/dev.to/pull/7294/files#diff-414ac17a58f5d19e6d0d81e26c3ff5d5R81-R96) is no longer shown.

nickytonline commented on app/javascript/onboarding/components/IntroSlide.jsx:
```diff
@@ -2,27 +2,98 @@ import { h, Component } from 'preact';
+            <div
+              className="terms-and-conditions-content"
+              /* eslint-disable react/no-danger */
+              dangerouslySetInnerHTML={{ __html: text }}
```
I imagine you're doing this as the ToS is being pulled as markup from the DB? If so, this is a valid use case for using `dangerouslySetInnerHTML `.

vaidehijoshi replied:
The ToS and CoC are rendered server-side in [this view](https://github.com/thepracticaldev/dev.to/blob/master/app/views/onboardings/show.html.erb#L19-L25) via Rails. To be honest, I just moved this from one file to another :)

nickytonline commented on app/javascript/onboarding/components/IntroSlide.jsx:
```diff
@@ -55,3 +184,5 @@ IntroSlide.propTypes = {
 
 export default IntroSlide;
+
+/* eslint-enable camelcase */
```
Instead of disabling and enabling the eslint camelCase rule, why not just rename the state properties `checked_code_of_conduct`, `checked_terms_and_conditions`. Maybe I'm missing something.

vaidehijoshi replied:
This code was preexisting but I believe the reason for this is because we use the attributes on the `state` to update the user, like we do [here](https://github.com/thepracticaldev/dev.to/pull/7294/files#diff-d16a96f0524905d872d0a314d8b10189R34). The properties are camel cased on the Rails side.

vaidehijoshi replied:
I guess the thing is that we _could_ rename the properties but then we have to construct a new object when we update the user on the backend with keys that are snakecased (because that's what Rails expects), and then we are renaming the properties to camel case only to map them back to their snakecased versions again 🤷

nickytonline replied:
In that case, leave it as is. 😄

nickytonline commented on app/javascript/onboarding/components/Navigation.jsx:
```diff
@@ -1,16 +1,36 @@
       )}
       {!hideNext && (
-        <button onClick={next} className="next-button" type="button">
+        <button
```
Consider using the [design system <Button />](https://devto-storybook.now.sh/?path=/story/components-buttons--description) component anywhere `<button />` is currently being used.

vaidehijoshi replied:
If it's okay with you, I'm thinking we can do this in a separate PR as well. There are a lot of slides in the onboarding flow that could be updated with the new design system button...and I like the idea of just updating the variables, buttons, etc in one fell swoop. I can add a card to [our project board](https://github.com/thepracticaldev/dev.to/projects/11#column-8471698) for this so that we make sure it gets done in #cooldown !

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -68,7 +69,11 @@ class ProfileForm extends Component {
+              <img
+                className="current-user-avatar"
+                alt="profile"
+                src={profile_image_90}
```
Doing the following will remove the need for disabling camel case completely in the file.

```suggestion
                src={profile_image_90  /* eslint-disable-line camelcase */}
```

vaidehijoshi replied:
So I tried doing this locally before committing the change...but it feels like it clutters up the file a lot more when I disable linting on a line-by-line basis. The JSX and functions feel hard to read with so many `eslint-disable-line`s everywhere instead of just disabling and enabling just once 🤷

nickytonline replied:
Ahh, I thought it was only in a couple of places. In that case, leave as is.

vaidehijoshi commented:
@nickytonline I addressed your suggestions in the latest commit, and then replied to the comments that I didn't address with some explanations of why 😆 let me know what you think!
